File: 1507-check-if-there-is-a-valid-path-in-a-grid/check-if-there-is-a-valid-path-in-a-grid.java
import java.util.*;

class Solution {

    // Directions: up, down, left, right
    int[][] dirs = {
        {-1,0},{1,0},{0,-1},{0,1}
    };

    // For each street type, allowed directions
    int[][][] roads = {
        {},
        {{0,-1},{0,1}},           //1 left-right
        {{-1,0},{1,0}},           //2 up-down
        {{0,-1},{1,0}},           //3 left-down
        {{0,1},{1,0}},            //4 right-down
        {{0,-1},{-1,0}},          //5 left-up
        {{0,1},{-1,0}}            //6 right-up
    };

    public boolean hasValidPath(int[][] grid) {

        int m = grid.length, n = grid[0].length;

        boolean[][] vis = new boolean[m][n];

        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0,0});
        vis[0][0]=true;

        while(!q.isEmpty()){

            int[] cur=q.poll();
            int r=cur[0], c=cur[1];

            if(r==m-1 && c==n-1)
                return true;

            for(int[] d : roads[grid[r][c]]){

                int nr=r+d[0];
                int nc=c+d[1];

                if(nr<0 || nc<0 || nr>=m || nc>=n || vis[nr][nc])
                    continue;

                // check neighbor connects back
                for(int[] back : roads[grid[nr][nc]]){

                    if(nr+back[0]==r && nc+back[1]==c){
                        vis[nr][nc]=true;
                        q.offer(new int[]{nr,nc});
                        break;
                    }
                }
            }
        }

        return false;
    }
}

File: 1507-check-if-there-is-a-valid-path-in-a-grid/check-if-there-is-a-valid-path-in-a-grid.py
class Solution:
    def hasValidPath(self, grid):

        dirs = {
            1: [(0,-1),(0,1)],
            2: [(-1,0),(1,0)],
            3: [(0,-1),(1,0)],
            4: [(0,1),(1,0)],
            5: [(0,-1),(-1,0)],
            6: [(0,1),(-1,0)]
        }

        m,n=len(grid),len(grid[0])

        vis=[[False]*n for _ in range(m)]

        q=deque([(0,0)])
        vis[0][0]=True

        while q:

            r,c=q.popleft()

            if r==m-1 and c==n-1:
                return True

            for dr,dc in dirs[grid[r][c]]:

                nr,nc=r+dr,c+dc

                if not(0<=nr<m and 0<=nc<n) or vis[nr][nc]:
                    continue

                for br,bc in dirs[grid[nr][nc]]:
                    if nr+br==r and nc+bc==c:
                        vis[nr][nc]=True
                        q.append((nr,nc))
                        break

        return False